File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return rec(root,p,q);
    }
    public TreeNode rec(TreeNode curr, TreeNode p, TreeNode q){
        if(curr == null){
            return null;
        }
        if(curr == p || curr == q){
            return curr;
        }
        TreeNode left = rec(curr.left,p,q);
        TreeNode right = rec(curr.right,p,q);
        if(left != null && right != null){
            return curr;
        }
        else if(left!= null){
            return left;
        }
        else if(right != null){
            return right;
        }
        return null;
        
    }
}

File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def LCA(node):
            if node.val == p.val or node.val == q.val:
                return node

            left = None
            right = None
            if node.left:
                left = LCA(node.left)
            if node.right:
                right = LCA(node.right)
            if left and right:
                return node
            if left or right:
                if left:
                    return left
                return right
            return None
        return LCA(root)